File: app/src/main/java/at/shockbytes/corey/ui/fragment/ScheduleFragment.java
package at.shockbytes.corey.ui.fragment;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.Toast;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import at.shockbytes.corey.R;
import at.shockbytes.corey.adapter.DaysScheduleAdapter;
import at.shockbytes.corey.adapter.ScheduleAdapter;
import at.shockbytes.corey.dagger.AppComponent;
import at.shockbytes.corey.storage.live.LiveScheduleUpdateListener;
import at.shockbytes.corey.ui.fragment.dialogs.InsertScheduleDialogFragment;
import at.shockbytes.corey.util.schedule.ScheduleItem;
import at.shockbytes.corey.util.schedule.ScheduleManager;
import at.shockbytes.util.AppUtils;
import at.shockbytes.util.adapter.BaseItemTouchHelper;
import at.shockbytes.util.view.EqualSpaceItemDecoration;
import butterknife.BindView;
import io.reactivex.functions.Consumer;

/**
 * @author Martin Macheiner
 *         Date: 26.10.2015.
 */
public class ScheduleFragment extends BaseFragment implements LiveScheduleUpdateListener,
        ScheduleAdapter.OnItemMoveListener, ScheduleAdapter.OnItemClickListener,
        InsertScheduleDialogFragment.OnScheduleItemSelectedListener {


    public static ScheduleFragment newInstance() {
        ScheduleFragment fragment = new ScheduleFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private ItemTouchHelper touchHelper;

    private ScheduleAdapter adapter;

    @Inject
    protected ScheduleManager scheduleManager;

    @BindView(R.id.fragment_schedule_rv)
    protected RecyclerView recyclerView;

    @BindView(R.id.fragment_schedule_rv_days)
    protected RecyclerView recyclerViewDays;

    @Override
    public void onStop() {
        super.onStop();
        scheduleManager.unregisterLiveForScheduleUpdates();
    }

    @Override
    public void onScheduleItemAdded(ScheduleItem item) {
        adapter.insertEntity(item);
    }

    @Override
    public void onScheduleItemDeleted(ScheduleItem item) {
        adapter.resetEntity(item);
    }

    @Override
    public void onScheduleItemChanged(ScheduleItem item) {
        adapter.updateEntity(item);
    }

    @Override
    public void onItemMove(ScheduleItem item, int from, int to) {

    }

    @Override
    public void onItemMoveFinished() {

        List<ScheduleItem> items = adapter.getScheduleData();
        for (ScheduleItem si : items) {
            scheduleManager.updateScheduleItem(si);
        }
    }

    @Override
    public void onItemDismissed(ScheduleItem item, int position) {

        if (!item.isEmpty()) {
            scheduleManager.deleteScheduleItem(item);
        }
        //adapter.addEntity(position, new ScheduleItem("", position));
        //recyclerView.invalidate();
    }

    @Override
    public void onItemClick(ScheduleItem item, View v, int position) {

        if (item.isEmpty()) {
            InsertScheduleDialogFragment fragment = InsertScheduleDialogFragment
                    .newInstance(position);
            fragment.setOnScheduleItemSelectedListener(this);
            fragment.show(getFragmentManager(), fragment.getTag());
        }
    }

    @Override
    public void onScheduleItemSelected(String item, int day) {
        scheduleManager.insertScheduleItem(new ScheduleItem(item, day));
    }

    @Override
    public void setupViews() {

        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), getColsForOrientation()));
        recyclerViewDays.setLayoutManager(new GridLayoutManager(getContext(), getColsForOrientation()));
        recyclerViewDays.setAdapter(new DaysScheduleAdapter(getContext(),
                Arrays.asList(getResources().getStringArray(R.array.days))));
        recyclerViewDays.addItemDecoration(new EqualSpaceItemDecoration(
                AppUtils.INSTANCE.convertDpInPixel(4, getContext())));

        adapter = new ScheduleAdapter(getContext(), new ArrayList<ScheduleItem>());
        recyclerView.setAdapter(adapter);
        recyclerView.setNestedScrollingEnabled(false);
        recyclerView.addItemDecoration(new EqualSpaceItemDecoration(
                AppUtils.INSTANCE.convertDpInPixel(4, getContext())));
        ItemTouchHelper.Callback callback = new BaseItemTouchHelper(adapter, true, BaseItemTouchHelper.DragAccess.VERTICAL);
        adapter.setOnItemMoveListener(this);
        adapter.setOnItemClickListener(this);

        touchHelper = new ItemTouchHelper(callback);
        touchHelper.attachToRecyclerView(recyclerView);

        scheduleManager.getSchedule().subscribe(new Consumer<List<ScheduleItem>>() {

            @Override
            public void accept(List<ScheduleItem> scheduleItems) {
                adapter.setData(scheduleItems);
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) {
                Toast.makeText(getContext(), throwable.toString(), Toast.LENGTH_LONG).show();
                throwable.printStackTrace();
            }
        });

        scheduleManager.registerLiveForScheduleUpdates(this);
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_schedule;
    }

    @Override
    protected void injectToGraph(@NotNull AppComponent appComponent) {
        appComponent.inject(this);
    }

    private int getColsForOrientation() {
        return getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT
                ? 1
                : 7;
    }

}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/ScheduleFragment.kt
package at.shockbytes.corey.ui.fragment

import android.content.res.Configuration
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.widget.Toast
import at.shockbytes.corey.R
import at.shockbytes.corey.adapter.DaysScheduleAdapter
import at.shockbytes.corey.adapter.ScheduleAdapter
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.schedule.ScheduleItem
import at.shockbytes.corey.schedule.ScheduleManager
import at.shockbytes.corey.storage.live.LiveScheduleUpdateListener
import at.shockbytes.corey.ui.fragment.dialog.InsertScheduleDialogFragment
import at.shockbytes.util.AppUtils
import at.shockbytes.util.adapter.BaseAdapter
import at.shockbytes.util.adapter.BaseItemTouchHelper
import at.shockbytes.util.view.EqualSpaceItemDecoration
import kotterknife.bindView
import javax.inject.Inject

/**
 * @author  Martin Macheiner
 * Date:    26.10.2015.
 */
class ScheduleFragment : BaseFragment(), LiveScheduleUpdateListener,
        BaseAdapter.OnItemMoveListener<ScheduleItem> {

    @Inject
    protected lateinit var scheduleManager: ScheduleManager

    private lateinit var touchHelper: ItemTouchHelper
    private lateinit var adapter: ScheduleAdapter

    private val recyclerView: RecyclerView by bindView(R.id.fragment_schedule_rv)
    private val recyclerViewDays: RecyclerView by bindView(R.id.fragment_schedule_rv_days)

    private val recyclerViewLayoutManager: RecyclerView.LayoutManager
        get() = if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        } else {
            GridLayoutManager(context, ScheduleAdapter.MAX_SCHEDULES)
        }

    override val layoutId = R.layout.fragment_schedule

    override fun onStart() {
        super.onStart()
        scheduleManager.registerLiveForScheduleUpdates(this)
    }

    override fun onStop() {
        super.onStop()
        scheduleManager.unregisterLiveForScheduleUpdates()
    }

    override fun onScheduleItemAdded(item: ScheduleItem) {
        adapter.insertScheduleItem(item)
    }

    override fun onScheduleItemDeleted(item: ScheduleItem) {
        adapter.resetEntity(item)
    }

    override fun onScheduleItemChanged(item: ScheduleItem) {
        adapter.updateScheduleItem(item)
    }

    override fun onItemMove(t: ScheduleItem, from: Int, to: Int) {
    }

    override fun onItemMoveFinished() {
        adapter.reorderAfterMove().forEach { scheduleManager.updateScheduleItem(it) }
    }

    override fun onItemDismissed(t: ScheduleItem, position: Int) {
        if (!t.isEmpty) {
            scheduleManager.deleteScheduleItem(t)
        }
    }

    override fun setupViews() {

        recyclerViewDays.layoutManager = recyclerViewLayoutManager
        recyclerViewDays.adapter = DaysScheduleAdapter(context!!, resources.getStringArray(R.array.days).toList())
        recyclerViewDays.addItemDecoration(EqualSpaceItemDecoration(AppUtils.convertDpInPixel(4, context!!)))

        recyclerView.layoutManager = recyclerViewLayoutManager
        adapter = ScheduleAdapter(context!!, listOf())
        recyclerView.adapter = adapter
        recyclerView.isNestedScrollingEnabled = false
        recyclerView.addItemDecoration(EqualSpaceItemDecoration(AppUtils.convertDpInPixel(4, context!!)))
        val callback = BaseItemTouchHelper(adapter, true, BaseItemTouchHelper.DragAccess.ALL)
        adapter.onItemMoveListener = this
        adapter.setOnScheduleItemSelectedListener { item, _, position ->
            onScheduleItemClicked(item, position)
        }

        touchHelper = ItemTouchHelper(callback)
        touchHelper.attachToRecyclerView(recyclerView)

        scheduleManager.schedule.subscribe({ scheduleItems ->
            adapter.data = scheduleItems.toMutableList()
        }, { throwable ->
            throwable.printStackTrace()
            Toast.makeText(context, throwable.toString(), Toast.LENGTH_LONG).show()
        })

    }

    override fun injectToGraph(appComponent: AppComponent) {
        appComponent.inject(this)
    }

    private fun onScheduleItemClicked(item: ScheduleItem, position: Int) {
        if (item.isEmpty) {
            InsertScheduleDialogFragment.newInstance()
                    .setOnScheduleItemSelectedListener { i ->
                        scheduleManager.insertScheduleItem(ScheduleItem(i, position))
                    }
                    .show(fragmentManager, "dialogfragment-insert-schedule")
        }
    }

    companion object {

        fun newInstance(): ScheduleFragment {
            val fragment = ScheduleFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }

}
